index.js

```diff
@@ -22,83 +22,384 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
-        throw new Error('Invalid I2C baud rate. Baud rates must be a multiple of 1000');
+  constructor(options) {
+    options = options || {};
+    super(options.pins || [ 'SDA0', 'SCL0' ]);
+
+    Object.defineProperties(this, {
+      [devices]: {
+        writable: true,
+        value: []
       }
-      sh.run('gpio load i2c ' + baudRate);
-    } else {
-      sh.run('gpio load i2c'); // Is this still necessary?
-    }
+    });
+
+    execSync('modprobe i2c-dev');
   }
 
   destroy() {
+    this[devices].forEach(function (device) {
+      device.closeSync();
+    });
+
+    this[devices] = [];
+
     super.destroy();
   }
 
+  [checkAlive]() {
+    if (!this.alive) {
+      throw new Error('Attempted to access a destroyed I2C peripheral');
+    }
+  }
+
+  [getDevice](address) {
+    var device = this[devices][address];
+
+    if (device === undefined) {
+      device = i2c.openSync(getBoardRevision() === VERSION_1_MODEL_B_REV_1 ? 0 : 1);
+      this[devices][address] = device;
+    }
+
+    return device;
+  }
+
   // function cb(err, buffer), returns undefined, register is optional
+  // Required by J5
   i2cRead(address, register, length, cb) {
-    throw new Error('Not implemented');
+    this[checkAlive]();
+
+    if (arguments.length === 3) {
+      cb = length;
+      length = register;
+      register = undefined;
+    }
+
+    checkAddress(address);
+    checkRegister(register);
+    checkLength(length);
+    checkCallback(cb);
+
+    var buffer = new Buffer(length);
+    function callback(err) {
+      if (err) {
+        return cb(err);
+      }
+      cb(null, buffer);
+    };
+
+    if (register === undefined) {
+      this[getDevice](address).i2cRead(address, length, buffer, callback);
+    } else {
+      this[getDevice](address).readI2cBlock(address, register, length, buffer, callback);
+    }
   }
 
   // Returns a buffer, register is optional
   i2cReadSync(address, register, length) {
-    throw new Error('Not implemented');
+    this[checkAlive]();
+
+    if (arguments.length === 2) {
+      length = register;
+      register = undefined;
+    }
+
+    checkAddress(address);
+    checkRegister(register);
+    checkLength(length);
+
+    var buffer = new Buffer(length);
+
+    if (register === undefined) {
+      this[getDevice](address).i2cReadSync(address, length, buffer);
+    } else {
+      this[getDevice](address).readI2cBlockSync(address, register, length, buffer);
+    }
+
+    return buffer;
   }
 
   // function cb(err, value), returns undefined, register is optional
   readByte(address, register, cb) {
-    throw new Error('Not implemented');
+    this[checkAlive]();
+
+    if (arguments.length === 2) {
+      cb = register;
+      register = undefined;
+    }
+
+    checkAddress(address);
+    checkRegister(register);
+    checkCallback(cb);
+
+    if (register === undefined) {
+      var buffer = new Buffer(1);
+      this[getDevice](address).i2cRead(address, buffer.length, buffer, function (err) {
```

Comment:
Sorry, I missed this before. This should be an arrow function. The code style for Raspi.js is to only use function declarations and arrow functions, not function expressions.